src/my-bot.ts

```diff
@@ -1,38 +1,423 @@
+                patrol: true,
+                submarine: true,
+                destroyer: true,
+                battleship: true,
+                carrier: true
+            },
+            hitmap: []
+        });
+        let shipPlaces: ShipPlace[] = [];
+        let done: boolean = false;
+        counter = 0;
+        do {
+            shipPlaces = [];
+            counter++;
+            shipPlaces[0] = this.getShipPlace(shipPlaces,5);
+            if (shipPlaces[0].StartingSquare.Column === -1) {
+                continue;
+            }
+            shipPlaces[1] = this.getShipPlace(shipPlaces,4);
+            if (shipPlaces[1].StartingSquare.Column === -1) {
+                continue;
+            }
+            shipPlaces[2] = this.getShipPlace(shipPlaces,3);
+            if (shipPlaces[2].StartingSquare.Column === -1) {
+                continue;
+            }
+            shipPlaces[3] = this.getShipPlace(shipPlaces,3);
+            if (shipPlaces[3].StartingSquare.Column === -1) {
+                continue;
+            }
+            shipPlaces[4] = this.getShipPlace(shipPlaces,2);
+            if (shipPlaces[4].StartingSquare.Column === -1) {
+                continue;
+            }
+            done = true;
+        } while ((!done) && (counter < 10000))
+        if (counter === 10000) {
+            throw new Error("Infinite loop generating ship positions");
         }
-        return { Row: "A", Column: 1 };  
+        return shipPlaces;
     }
 
-    private getNextTarget(position) {
+    public selectTarget(gamestate): Promise<Position> {
+        return new Promise((resolve, reject) => {
+            firebase.database().ref('matches/' + this.matchId.toString()).once('value').then((snapshot) => { 
+                let snapCopy = snapshot.val();
+                if (gamestate.MyShots.length === 0) {
+                    resolve(this.randomShot(gamestate));
+                }
+                const previousShot: Shot = gamestate.MyShots && gamestate.MyShots[gamestate.MyShots.length-1];
+                let target: Position = { Row: 'A', Column: 1 };
+                if (previousShot.WasHit) {
+                    snapCopy.hitmode = true;
+                    if (snapCopy.hitmap) {
+                        snapCopy.hitmap.push(previousShot.Position);
+                    } else {
+                        snapCopy.hitmap = [previousShot.Position];
+                    }
+                }
+                firebase.database().ref('matches/' + this.matchId.toString()).set(snapCopy);
+                if (snapCopy.hitmode) {
+                    target = this.track(gamestate,snapCopy);
+                } else if (snapCopy.hitmap) {
+                    target = this.randomShot(gamestate,snapCopy.hitmap);
+                } else {
+                    target = this.randomShot(gamestate);
+                }
+                return resolve(target);
+            })
+            .catch((err) => {console.log(err)});

```

Comment:
Neat, but I'm a little wary of this simply logging and hiding the error. What errors are you expecting - firebase ones? Wouldn't it be better to rethrow them so that your error handling on the router catches them?